src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs

```diff
@@ -788,5 +789,12 @@ public void ResetLoadingStatusIndicator()
         {
             _loadingStatusIndicator.Reset(string.Empty);
         }
+
+        private void InfiniteScrollListUnloaded(object sender, RoutedEventArgs e)
+        {
+            Items.Clear();
```

Comment:
If you comment out this line, do you still see memory improvements?

Reply:
Yes, I do. I tested several times with and without my changes.

Reply:
@donnie-msft Do you know why we have this memory leak?

Reply:
It would take some investigation. how much of this line (what %) of the improvement comes from here?
If you don't clear items, do you see a ~10% improvement with your other changes?
If you do clear items, do you see a ~70% improvement?
That will tell us where we should focus more. 

My guess is the `Items` may have an issue with one of the child controls that hooks up events, or similar, and those are causing leaks. If you find a large portion of improvement here, then we should dig into the XAML child controls in https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageItemControl.xaml.

One idea that I just had was... what if you pull a very old NuGet.Client (~1-2 years ago) before we added Deprecation. You could quickly pinpoint whether we have been improving or getting worse with these leaks. Maybe Deprecation work in 2019 is the focus area. Just a thought.

Reply:
@donnie-msft I don't have 2 separate data for Items.Clear() and [Model.CachedUpdates = null;](https://github.com/NuGet/NuGet.Client/blob/ca8516edd8fe6c8a86d0833ce71de999d3f7a61a/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs#L904) changes.
But here are some data for `open and close PM UI 2 times`  with my 2 above changes:

`NugetVersion` before my change 4260 instances were in memory  and keep growing high rate after each I open PM UI.
After my change it was around 80 instances  and kept growing slow rate after repeated open/close of PM UI.

`VersionRange` before my change about 3491 instances were in memory and keep growing after each I open PM UI.
After my change always hover around 20 instances after repeated open/close PM UI.

Can you tell me which month of 2019 was it (Deprecation work)?

Reply:
I think this is the initial Deprecation PR https://github.com/NuGet/NuGet.Client/pull/2899
June 2019. I found it by [searching closed PRs](https://github.com/NuGet/NuGet.Client/pulls?q=is%3Apr+is%3Aclosed+deprecation).

Reply:
@donnie-msft I reverted to around that time. But was not able to build the code, so many cryptic errors.

Reply:
@donnie-msft I found something, `PackageManagerControl` is not properly released [here](https://github.com/NuGet/NuGet.Client/blob/ca8516edd8fe6c8a86d0833ce71de999d3f7a61a/src/NuGet.Clients/NuGet.PackageManagement.UI/PackageManagerWindowPane.cs#L44).

![image](https://user-images.githubusercontent.com/8766776/87212957-4682f980-c2d6-11ea-8729-a3cf229804ef.png)

Reply:
Is that window pane closed? Maybe it's staying cached somehow by VS so it can be quickly reused.... But then we never reuse it and create a new one?

I would imagine the variable there for the control gets GC as usual. You could try setting it to null in the Cleanup to see what that does.
Thoughts @rrelyea ?

Reply:
@donnie-msft Can you give me code suggestion? I don't get which Cleanup()? Currently it's already inside one Cleanup(), there are other one but not sure how to set null [there](CleanUp).

Reply:
> @donnie-msft I found something, `PackageManagerControl` is not properly released [here](https://github.com/NuGet/NuGet.Client/blob/ca8516edd8fe6c8a86d0833ce71de999d3f7a61a/src/NuGet.Clients/NuGet.PackageManagement.UI/PackageManagerWindowPane.cs#L44). Do you know what to do this case?
> 
> ![image](https://user-images.githubusercontent.com/8766776/87212957-4682f980-c2d6-11ea-8729-a3cf229804ef.png)
> 

This file has a CleanUp() method but it doesn't do anything with the PackageManagerControl that you mentioned. I don't know exact code change to make, but try setting it to null in this method. 

I'm guessing we will need to dive into the VS guidance for handling window panes. As I said, I think probably we are caching our control forever, and we never reuse that cached control. That's a guess based on reading your findings

Reply:
Sorry, it does already set to null. I would spend a little time debugging to see what the dispose and other events in this file are doing. Maybe null is not the right thing. Maybe there's another reference to this control somewhere else?

Reply:
@donnie-msft There are so many references to this control on memory dump. Not sure which one actually causing problem. Maybe they're implicit indication of something I don't know yet. Whom do we ask for guidance?

Reply:
This `PackageManagerControl` created here: https://github.com/NuGet/NuGet.Client/blob/3328c404204c714787d55d736833042f53212d3c/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L740
Once created it's passed to VS shell.

Reply:
@mirkomuggli Recently made a change in the codespaces feature branch, that changes how we open the PM UI, I'm wondering if that makes any difference.

Reply:
@rrelyea @srdjanjovcic @donnie-msft Thank you. With their help I was able to find root cause of problem. It was `EllipseDetails` behaving badly. 
![image](https://user-images.githubusercontent.com/8766776/87361997-568f1900-c522-11ea-959b-0386eb47ca12.png)

Reply:
https://stackoverflow.com/questions/18542940/can-bindings-create-memory-leaks-in-wpf